kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java

```diff
@@ -0,0 +1,218 @@
+ */
+public class KCBQThreadPoolExecutor extends ThreadPoolExecutor {
+
+  private static final Logger logger = LoggerFactory.getLogger(KCBQThreadPoolExecutor.class);
+
+  private ConcurrentHashMap<String, AtomicInteger> activeThreadTopicCount =
+      new ConcurrentHashMap<>();
+  private ConcurrentSet<Throwable> encounteredErrors = new ConcurrentSet<>();
+  private TopicPartitionManager topicPartitionManager;
+  private int topicThreadLimit;
+
+  /**
+   * @param config the {@link BigQuerySinkTaskConfig}
+   * @param context the {@link SinkTaskContext}
+   * @param workQueue the queue for storing tasks.
+   */
+  public KCBQThreadPoolExecutor(BigQuerySinkTaskConfig config,
+                                SinkTaskContext context,
+                                BlockingQueue<Runnable> workQueue) {
+    super(config.getInt(BigQuerySinkTaskConfig.THREAD_POOL_SIZE_CONFIG),
+          config.getInt(BigQuerySinkTaskConfig.THREAD_POOL_SIZE_CONFIG),
+          // the following line is irrelevant because the core and max thread counts are the same.
+          1, TimeUnit.SECONDS,
+          workQueue);
+    topicPartitionManager = new TopicPartitionManager(context);
+    topicThreadLimit = config.getInt(BigQuerySinkTaskConfig.TOPIC_MAX_THREADS_CONFIG);
+  }
+
+  @Override
+  protected void beforeExecute(Thread thread, Runnable runnable) {
+    super.beforeExecute(thread, runnable);
+
+    if (runnable instanceof TableWriter) {
+      TableWriter tableWriter = (TableWriter) runnable;
+      String topic = tableWriter.getTopic();
+      activeThreadTopicCount.putIfAbsent(topic, new AtomicInteger(0));
+
+      int topicThreadCount = activeThreadTopicCount.get(topic).incrementAndGet();
+      if (topicThreadCount > topicThreadLimit) {
+        topicPartitionManager.pause(topic, topicThreadCount);
+      }
+    }
+  }
+
+  @Override
+  protected void afterExecute(Runnable runnable, Throwable throwable) {
+    super.afterExecute(runnable, throwable);
+
+    if (runnable instanceof TableWriter) {
+      TableWriter tableWriter = (TableWriter) runnable;
+      String topic = tableWriter.getTopic();
+
+      int topicThreadCount = activeThreadTopicCount.get(topic).decrementAndGet();
+      if (topicThreadCount < topicThreadLimit) {
+        topicPartitionManager.resume(topic);
+      }
+    }
+
+    if (throwable != null) {
+      logger.error("Task failed with {} error: {}",
+                   throwable.getClass().getName(),
+                   throwable.getMessage());
+      encounteredErrors.add(throwable);
+    }
+  }
+
+  /**
+   * Wait for all the currently queued tasks to complete, and then return.
+   *
+   * @throws BigQueryConnectException if any of the tasks failed.
+   * @throws InterruptedException if interrupted while waiting.
+   */
+  public void awaitCurrentTasks() throws InterruptedException, BigQueryConnectException {
+    int maximumPoolSize = getMaximumPoolSize();
+    CountDownLatch countDownLatch = new CountDownLatch(maximumPoolSize);
```

Comment:
I don't quite follow this code. I think the trick that you're playing is that you're stuffing the max pull size number of countdown latches into the thread pool and awaiting on it to make sure that everything prior to it has finished.

What was the reason for doing this vs. calling join() on each remaining thread in the pool directly?